README.md
simplex
=======

My implementation of the Nelder-Mead Simplex algorithm for unconstrained nonlinear programming.

Pull requests, comments, and suggestions are welcomed!

Installation
============
I'm assuming you are wise and you're using virtualenv and virtualenvwrapper. If not, go and [install them now](http://virtualenvwrapper.readthedocs.org/en/latest/).

In order to install the package, run in the terminal:

``` console
$ python setup.py sdist
```

Then:

``` console
$ pip install dist/Simplex-0.1.0.tar.gz
```

And you're done!

Basic usage
===========

A typical usage would look as follows:

``` python
from simplex.algorithm import NelderMeadSimplex

import numpy as np

# Define objective function
def objective(xs):
    x1, x2 = xs[0], xs[1]
    return x1**2 + x2**2 - 3*x1 - x1*x2 + 3

# Initial simplex
simplex = np.array([[0,0], [0,1], [1,0]], dtype=np.float)

# Initialise NelderMead simplex algorithm
nm = NelderMeadSimplex(objective, simplex)

# Minimise the objective function
solution = nm.solve()

print("Minimum at {}".format(solution))
```

More examples
=============
You can find more examples in the ```examples/``` folder.

License
=======

License information can be found in License.txt.


CHANGES.txt
v0.1.0, 22-10-2013 -- Initial release.


main.py
from simplex import NelderMead

import numpy as np
import matplotlib.pyplot as plt

# Define objective function
def objective(xs):
    x1, x2 = xs[0], xs[1]
    return x1**2 + x2**2 - 3*x1 - x1*x2 + 3

simplices = []
def callback(args):
    simplices.append(args[0])

# Initial simplex
simplex = np.array([[0,0], [0,1], [1,0]], dtype=np.float)

# Initialise NelderMead simplex algorithm
nm = NelderMead(objective, simplex, epsilon=1e-1, callback=callback)

# Minimise the objective function
solution = nm.solve()

print("Minimum at {}".format(solution))

# Plot
# Tabulate objective function
x = np.linspace(-1, 3, 1000)
y = np.linspace(-2, 2, 1000)
X, Y = np.meshgrid(x, y)
Z = X**2 + Y**2 - 3*X - X*Y + 3

plt.figure()
cs = plt.contour(X, Y, Z, 25)
plt.clabel(cs, inline=1, fontsize=10)

for simplex in simplices:

    lines = []
    for i in range(3):
        for j in range(i, 3):
            if j == i:
                continue

            plt.plot(*zip(simplex[i], simplex[j]), c='black')

plt.grid()
plt.savefig('simplex.pdf')


simplex.py
import logging

import numpy as np


class NelderMead:
    def __init__(self, objective, simplex, epsilon=1e-6, callback=None):
        """
        Arguments:
        objective -- objective function to be minimised. It is
        assumed the function takes a numpy array (vector) as input.
        simplex -- initial simplex. It is assumed the simplex is
        a 2D numpy array.
        epsilon -- (optional) 
        """
        self.objective = objective
        self.simplex = simplex
        self.alpha = 1
        self.beta = 0.5
        self.gamma = 2
        self.epsilon = epsilon
        self.__callback = callback

    def solve(self):
        """
        Returns the point (numpy array) that minimises
        the objective function.
        """
        # Initialise numpy arrays
        n = self.simplex.shape[0] - 1
        func_simplex = np.empty(n+1, dtype=np.float)
        func_simplex_without_1 = np.empty(n, dtype=np.float)

        while True:
            # Call back
            self.call_back()

            # Compute function values for the current simplex
            for i in np.arange(n+1):
                func_simplex[i] = self.objective(self.simplex[i])

            # Compute x for which the objective assumes the highest value
            high_index = np.argmax(func_simplex)
            high = self.simplex[high_index]
            f_high = self.objective(high)

            # Compute function values for the simplex without the highest
            # value
            j = 0
            for i in np.arange(n+1):
                value = self.simplex[i]

                if (value == high).all():
                    continue

                func_simplex_without_1[j] = self.objective(value)
                j += 1

            # Compute the centroid value
            centroid = self.__centroid(high)

            # Check for convergence to minimum
            centroids = np.ones(n+1, dtype=np.float) * self.objective(centroid)
            stopping_condition = np.sqrt(np.sum((func_simplex - centroids)**2) / (n+1))
            if stopping_condition < self.epsilon:
                break

            # Compute x for which the objective assumes the lowest value
            low = self.simplex[np.argmin(func_simplex)]
            f_low = self.objective(low)

            # Reflect
            reflection = self.__reflect(centroid, high)
            f_reflection = self.objective(reflection)

            if f_low > f_reflection:
                # Reflection is the new lowest value. Expand
                expansion = self.__expand(centroid, reflection)
                f_expansion = self.objective(expansion)

                if f_reflection > f_expansion:
                    # Expansion is the new highest value. Create new simplex
                    self.simplex[high_index] = expansion

                else:
                    # Expansion failed. Reflection is the new highest value.
                    # Create new simplex
                    self.simplex[high_index] = reflection

            elif np.amax(func_simplex_without_1, axis=0) >= f_reflection and f_reflection >= f_low:
                # Reflection is the new highest value. Create new simplex
                self.simplex[high_index] = reflection

            else:
                # Define augmented highest value
                if f_high > f_reflection:
                    high = reflection
                    f_high = self.objective(high)
                
                # Contract
                contraction = self.__contract(centroid, high)
                f_contraction = self.objective(contraction)

                if f_contraction <= f_high:
                    # Contraction is the new highest value.
                    # Create new simplex
                    self.simplex[high_index] = contraction

                else:
                    # Create new simplex
                    for i in np.arange(n+1):
                        self.simplex[i] = self.simplex[i] + 0.5 * (low - self.simplex[i])

        return self.simplex[np.argmin(func_simplex)]

    def __centroid(self, high):
        """
        Returns the centroid of all points in the simplex
        except the one that maximises the objective function.

        Arguments:
        high -- point that maximises the objective function
        """
        n = self.simplex.shape[0] - 1
        m = self.simplex.shape[1]
        values = np.empty((n,m), dtype=np.float)

        j = 0
        for i in np.arange(n+1):
            value = self.simplex[i]

            if (value == high).all():
                continue

            values[j] = value
            
            j += 1

        return 1 / n * np.sum(values, axis=0)

    def __reflect(self, centroid, value):
        """
        Returns point that correspond to the reflection
        step of the Nelder-Mead algorithm.

        Arguments:
        centroid -- the centroid
        value -- point to be reflected
        """
        return centroid + self.alpha * (centroid - value)

    def __expand(self, centroid, value):
        """
        Returns point that correspond to the expansion
        step of the Nelder-Mead algorithm.

        Arguments:
        centroid -- the centroid
        value -- point to be expanded
        """
        return centroid + self.gamma * (value - centroid)

    def __contract(self, centroid, value):
        """
        Returns point that correspond to the contraction
        step of the Nelder-Mead algorithm.

        Arguments:
        centroid -- the centroid
        value -- point to be contracted
        """
        return centroid + self.beta * (value - centroid)

    def call_back(self):
        if self.__callback:
            args = [np.copy(self.simplex).tolist()]
            self.__callback(args)


simplex/algorithm.py
import numpy as np


class NelderMeadSimplex:
    def __init__(self, objective, simplex, epsilon=1e-6, max_iter=1000, callback=None):
        """
        Arguments:
        objective -- objective function to be minimised. It is
        assumed the function takes a numpy array (vector) as input.
        simplex -- initial simplex. It is assumed the simplex is
        a 2D numpy array.
        epsilon -- (optional) convergence criterion
        max_iter -- (optional) maximum number of iterations
        callback -- (optional) callback function accepting array
        arguments
        """
        self.objective = objective
        self.simplex = simplex
        self.alpha = 1
        self.beta = 0.5
        self.gamma = 2
        self.epsilon = epsilon
        self.max_iter = max_iter
        self.__callback = callback

    def solve(self):
        """
        Returns the point (numpy array) that minimises
        the objective function.
        """
        # Initialise iteration counter
        it = 0

        # Initialise numpy arrays
        n = self.simplex.shape[0] - 1
        func_simplex = np.empty(n+1, dtype=np.float)
        func_simplex_without_1 = np.empty(n, dtype=np.float)

        while True and it < self.max_iter:
            # Call back if defined
            if self.__callback:
                args = [np.copy(self.simplex).tolist()]
                self.__callback(args)

            # Compute function values for the current simplex
            for i in np.arange(n+1):
                func_simplex[i] = self.objective(self.simplex[i])

            # Compute x for which the objective assumes the highest value
            high_index = np.argmax(func_simplex)
            high = self.simplex[high_index]
            f_high = self.objective(high)

            # Compute function values for the simplex without the highest
            # value
            j = 0
            for i in np.arange(n+1):
                value = self.simplex[i]

                if (value == high).all():
                    continue

                func_simplex_without_1[j] = self.objective(value)
                j += 1

            # Compute the centroid value
            centroid = self.__centroid(high)

            # Check for convergence to minimum
            centroids = np.ones(n+1, dtype=np.float) * self.objective(centroid)
            stopping_condition = np.sqrt(np.sum((func_simplex - centroids)**2) / (n+1))
            if stopping_condition < self.epsilon:
                break

            # Compute x for which the objective assumes the lowest value
            low = self.simplex[np.argmin(func_simplex)]
            f_low = self.objective(low)

            # Reflect
            reflection = self.__reflect(centroid, high)
            f_reflection = self.objective(reflection)

            if f_low > f_reflection:
                # Reflection is the new lowest value. Expand
                expansion = self.__expand(centroid, reflection)
                f_expansion = self.objective(expansion)

                if f_reflection > f_expansion:
                    # Expansion is the new highest value. Create new simplex
                    self.simplex[high_index] = expansion

                else:
                    # Expansion failed. Reflection is the new highest value.
                    # Create new simplex
                    self.simplex[high_index] = reflection

            elif np.amax(func_simplex_without_1, axis=0) >= f_reflection and f_reflection >= f_low:
                # Reflection is the new highest value. Create new simplex
                self.simplex[high_index] = reflection

            else:
                # Define augmented highest value
                if f_high > f_reflection:
                    high = reflection
                    f_high = self.objective(high)
                
                # Contract
                contraction = self.__contract(centroid, high)
                f_contraction = self.objective(contraction)

                if f_contraction <= f_high:
                    # Contraction is the new highest value.
                    # Create new simplex
                    self.simplex[high_index] = contraction

                else:
                    # Create new simplex
                    for i in np.arange(n+1):
                        self.simplex[i] = self.simplex[i] + 0.5 * (low - self.simplex[i])

            it += 1

        return self.simplex[np.argmin(func_simplex)]

    def __centroid(self, high):
        """
        Returns the centroid of all points in the simplex
        except the one that maximises the objective function.

        Arguments:
        high -- point that maximises the objective function
        """
        n = self.simplex.shape[0] - 1
        m = self.simplex.shape[1]
        values = np.empty((n,m), dtype=np.float)

        j = 0
        for i in np.arange(n+1):
            value = self.simplex[i]

            if (value == high).all():
                continue

            values[j] = value
            
            j += 1

        return 1 / n * np.sum(values, axis=0)

    def __reflect(self, centroid, value):
        """
        Returns point that correspond to the reflection
        step of the Nelder-Mead algorithm.

        Arguments:
        centroid -- the centroid
        value -- point to be reflected
        """
        return centroid + self.alpha * (centroid - value)

    def __expand(self, centroid, value):
        """
        Returns point that correspond to the expansion
        step of the Nelder-Mead algorithm.

        Arguments:
        centroid -- the centroid
        value -- point to be expanded
        """
        return centroid + self.gamma * (value - centroid)

    def __contract(self, centroid, value):
        """
        Returns point that correspond to the contraction
        step of the Nelder-Mead algorithm.

        Arguments:
        centroid -- the centroid
        value -- point to be contracted
        """
        return centroid + self.beta * (value - centroid)


examples/no-minimum.py
from simplex.algorithm import NelderMeadSimplex

import numpy as np
import matplotlib.pyplot as plt

from matplotlib import rc
rc('font',**{'family':'sans-serif','sans-serif':['Helvetica']})
## for Palatino and other serif fonts use:
#rc('font',**{'family':'serif','serif':['Palatino']})
rc('text', usetex=True)

# Define objective function
def objective(xs):
    x1, x2 = xs[0], xs[1]
    return x1*x2

# Define callback function
simplices = []
def callback(args):
    simplices.append(args[0])

# Initial simplex
simplex = np.array([[0,0], [1,1], [0,1]], dtype=np.float)

# Initialise NelderMead simplex algorithm
nm = NelderMeadSimplex(objective, simplex, epsilon=1e-6, max_iter=8, callback=callback)

# Minimise the objective function
solution = nm.solve()

print("Minimum at {}".format(solution))

# Tabulate objective function
x = np.linspace(-10, 10, 1000)
y = np.linspace(-10, 10, 1000)
X, Y = np.meshgrid(x, y)
Z = X*Y

# Plot function contours together with the evolution of
# the simplices as they approach the minimum
fig = plt.figure()
cs = plt.contour(X, Y, Z, 20)
plt.clabel(cs, inline=1, fontsize=10)

for simplex in simplices:

    lines = []
    for i in range(3):
        for j in range(i, 3):
            if j == i:
                continue

            plt.plot(*zip(simplex[i], simplex[j]), c='black')

plt.xlabel(r"$$x_1$$")
plt.ylabel(r"$$x_2$$")
plt.grid()
plt.savefig('no-minimum.png', bbox_inches=0, dpi=80)


examples/one-minimum.py
from simplex.algorithm import NelderMeadSimplex

import numpy as np
import matplotlib.pyplot as plt

from matplotlib import rc
rc('font',**{'family':'sans-serif','sans-serif':['Helvetica']})
## for Palatino and other serif fonts use:
#rc('font',**{'family':'serif','serif':['Palatino']})
rc('text', usetex=True)

# Define objective function
def objective(xs):
    x1, x2 = xs[0], xs[1]
    return x1**2 + x2**2 - 3*x1 - x1*x2 + 3

# Define callback function
simplices = []
def callback(args):
    simplices.append(args[0])

# Initial simplex
simplex = np.array([[0,0], [0,1], [1,0]], dtype=np.float)

# Initialise NelderMead simplex algorithm
nm = NelderMeadSimplex(objective, simplex, epsilon=1e-6, callback=callback)

# Minimise the objective function
solution = nm.solve()

print("Minimum at {}".format(solution))

# Tabulate objective function
x = np.linspace(-1, 3, 1000)
y = np.linspace(-2, 2, 1000)
X, Y = np.meshgrid(x, y)
Z = X**2 + Y**2 - 3*X - X*Y + 3

# Plot function contours together with the evolution of
# the simplices as they approach the minimum
plt.figure()
cs = plt.contour(X, Y, Z, 25)
plt.clabel(cs, inline=1, fontsize=10)

for simplex in simplices:

    lines = []
    for i in range(3):
        for j in range(i, 3):
            if j == i:
                continue

            plt.plot(*zip(simplex[i], simplex[j]), c='black')

plt.xlabel(r"$$x_1$$")
plt.ylabel(r"$$x_2$$")
plt.grid()
plt.savefig('one-minimum.png', bbox_inches=0, dpi=80)


examples/two-minima.py
from simplex.algorithm import NelderMeadSimplex

import numpy as np
import matplotlib.pyplot as plt

from matplotlib import rc
rc('font',**{'family':'sans-serif','sans-serif':['Helvetica']})
## for Palatino and other serif fonts use:
#rc('font',**{'family':'serif','serif':['Palatino']})
rc('text', usetex=True)

# Define objective function
def objective(xs):
    x1, x2 = xs[0], xs[1]
    return x1**4 + x2**4 - 1.25 * x1**2 + 0.25

# Define callback function
simplices = []
def callback(args):
    simplices.append(args[0])

# Initial simplex
# simplex = np.array([[-2,0], [-1,2], [-1.75,2]], dtype=np.float)
simplex = np.array([[2,0], [1,2], [1.75,2]], dtype=np.float)

# Initialise NelderMead simplex algorithm
nm = NelderMeadSimplex(objective, simplex, epsilon=1e-6, callback=callback)

# Minimise the objective function
solution = nm.solve()

print("Minimum at {}".format(solution))

# Tabulate objective function
x = np.linspace(-3, 3, 1000)
y = np.linspace(-3, 3, 1000)
X, Y = np.meshgrid(x, y)
Z = X**4 + Y**4 - 1.25 * X**2 + 0.25

# Plot function contours together with the evolution of
# the simplices as they approach the minimum
fig = plt.figure()
cs = plt.contour(X, Y, Z, 20)
plt.clabel(cs, inline=1, fontsize=10)

for simplex in simplices:

    lines = []
    for i in range(3):
        for j in range(i, 3):
            if j == i:
                continue

            plt.plot(*zip(simplex[i], simplex[j]), c='black')

plt.xlabel(r"$$x_1$$")
plt.ylabel(r"$$x_2$$")
plt.grid()
plt.savefig('two-minima.png', bbox_inches=0, dpi=80)
